### ISSUE TITLE
Enter event on form wrapped input should not have e.preventDefault()

### ISSUE THREAD

**Issue by hzhu** [paypal/downshift#330]
First of all thanks @kentcdodds for authoring this library! Also thanks to everyone else who's been helping Kent maintain this project ❤️.

* `downshift` version: 1.28.0
* `node` version: v8.1.3
* `yarn` version: 1.2.1

### What you did:
I type the string "black" (lowercase) into the form example in Downshift's storybook - then hit `ENTER`.

### What happened:
The form submission handler does not get called. However, my expectation is that the form submit handler gets called even though lowercase "black" isn't a part of the list. This is important when building autocomplete for search services. When a user hits `ENTER`, I want to still be able to execute a search on lowercase "black", even though it is not part of the list in the combobox:

<img width="572" alt="screen shot 2018-02-13 at 11 16 09 am" src="https://user-images.githubusercontent.com/1811365/36174480-4e1ed584-10c1-11e8-85a8-b99b629b610a.png">

### Reproduction repository:
Reproducible in Downshift Storybook in the [form example](https://github.com/paypal/downshift/blob/master/stories/examples/form.js).

### Problem description:
I think that a `<form>` submission handler should always be called on `ENTER`, even if the combo box list is open. I believe in the past, a `<form>` submission was not possible entirely, see #161. 

Pull request #161 made it possible to submit the `<form>`, but not when the combobox list `isOpen`.

### Suggested solution:
I am not sure I have an elegant solution. I am seeking the help of the community on this one. Please let me know if I got this all wrong and I'm simply not understanding the correct usage of Downshift.

A solution could be for Downshift to accept a prop `isForm` to Downshift. Then we will not [call `e.preventDefault`](https://github.com/paypal/downshift/blob/master/src/downshift.js#L506) when a user presses `ENTER` at all if the consumer of the component tells Downshift that a `<form>` wraps its `<input>`.

**Comment by kentcdodds** [comment]
Hey @hzhu,

Thanks for the issue! I'm not sure of the best solution for this to be honest. I can think of several things you could do, but I'm not super fond of them.

I don't think I want to add a prop especially for this case. I think we can come up with something else. I'd love people to give their input to solving this problem. Anyone have ideas?

**Comment by newyork-anthonyng** [comment]
@hzhu I was working on this [commit](https://github.com/newyork-anthonyng/downshift/commit/75377bf34f6981f018406756cfbffeab26751799) that seems to address the issue.

I'm not sure if it's the most elegant. I had to control the `isOpen` and `selectedItem` of the `Downshift` component. I also had to pass the `onSubmit` function to the `input` element.

Something else that I tried playing with but couldn't make it work:
The `Enter` keydown event is prevented only when the menu is open ([see code here](https://github.com/paypal/downshift/blob/master/src/downshift.js#L505)). I had an idea to close the menu first, and then redispatch the `Enter` keydown event.

```jsx
<Input
    {...getInputProps({
        placeholder: 'Enter color here',
        onKeyDown: e => {
            e.persist();
            // Use closeMenu action function that Downshift provides
            closeMenu(() => {
                // somehow redispatch the Synthetic event so that Downshift could take care of it
                // https://stackoverflow.com/questions/11974262/how-to-clone-or-re-dispatch-dom-events
                // this is where I failed 😟
            });      
        }
     })}
/>
```

**Comment by austintackaberry** [comment]
It seems like there are really only 2 options to let downshift handle the event correctly as the code stands right now. You either find a way to close the menu or you find a way to highlight the item before the Enter keyDownHandler is run.

@newyork-anthonyng for my app, I ended up doing something similar to your commit. I can't think of a better way to do it.

**Comment by austintackaberry** [comment]
I'm sure there are many reasons this is a terrible idea, but the implementation below allows the user to let downshift handle the event while still getting the functionality that @hzhu wanted.

https://codesandbox.io/s/x229880xjo

The general idea is
```
if (item === inputValue && highlightedIndex !== index) {
   // take control of highlightedIndex and force it to equal index
}
else {
   // give control of highlightedIndex back to Downshift
}
```

**Comment by newyork-anthonyng** [comment]
@austintackaberry I forked your [CodeSandbox example](https://codesandbox.io/s/xv8nk618xp) and wrapped it with a `<form />`. It looks like the `onSubmit` isn't firing.

Could you take a look at see if I'm missing something?

**Comment by kentcdodds** [comment]
What if we allowed you to do this:

```javascript
const handlers = {
  Enter(event) {
    if (this.getState().isOpen) {
      this.selectHighlightedItem({
        type: Downshift.stateChangeTypes.keyDownEnter,
      })
    }
  }
}

const ui = <Downshift keyDownHandlers={handlers}>
 {() => <div>your ui</div>}
</Downshift>
```

This would be fairly trivial to implement. We'd just need to add a little bit of logic [here](https://github.com/paypal/downshift/blob/83cb60beae498f48c454f9407e7a6bcf256879cb/src/downshift.js#L670-L674). Probably change it to this:

```javascript
input_handleKeyDown = event => {
  const handler = this.props.keyDownHandlers[event.key] || this.keyDownHandlers[event.key]
  if (handler) {
    handler.call(this, event)
  }
}
```

It would be an advanced feature and we should mention that in the docs to avoid using it because it exposes the downshift instance in a way that _could_ lead to your code breaking if we decide to change a method you're relying on...

Alternatively we could do: `handler.call(this.getStateAndHelpers(), event)` which would only allow you to access the same things you can access in the render prop... I think I like that better. It would require us to expose `moveHighlightedIndex`, but I'm cool with that.

Thoughts? Anyone wanna try to take this on? We'll need at least one or two tests and the `keyDownHandlers` prop will need prop types and a default value.

The more I think about this the more I like it as a solution. It's not the best, but it seems like a good way to get around this problem without complicating the codebase or the usage too much.

**Comment by austintackaberry** [comment]
Oops, I misread the original concern. Yeah, my [codesandbox](https://codesandbox.io/s/x229880xjo) doesn't submit the form (especially considering there is no form). All it does is select the item that equals inputValue when hitting Enter.

Why do you need to actually submit the form though? Why not just make your form a [controlled component](https://reactjs.org/docs/forms.html#controlled-components) and let React handle the state of each element in the form?

**Comment by hzhu** [comment]
> Why do you need to actually submit the form though? Why not just make your form a controlled component and let React handle the state of each element in the form?

Hi Austin, just to preface, I could be wrong as I'm pretty new to working with autocomplete UIs 🤐 . Some apps like Amazon.com implement their search autocomplete as a form. I think it would be standard behavior when a user hits `ENTER` that the `<form>`'s `onSubmit` handler is invoked. Note, the form doesn't need to be submitted to a server. The page doesn't need to refresh.

> Thoughts? Anyone wanna try to take this on? We'll need at least one or two tests and the keyDownHandlers prop will need prop types and a default value.

Hi Kent, I'd love to take this on as my first open source contribution to Downshift! I just tested your suggestion and it works. It seems non-invasive to the existing codebase. My teammate (who is helping me implement autocomplete at work using Downshift) has been working on a solution over the weekend. I'd love to see what he has in mind also!

**Comment by kentcdodds** [comment]
Thanks for that thorough overview @indiesquidge.

I think a lot of people forget that there are a _TON_ of use cases with user experience for components that use downshift. Sure YouTube, Google, and Amazon may have autocompletes that behave similarly (and differently from downshift's default), but that's because they actually have similar "search" kinds of use cases. What about an autocomplete where you can only select one of a given set of options? That was the original use case for downshift which is why that's the default. That said, you can support the experience that YouTube, Google, and Amazon give you with downshift without weird use-case-specific props or hacks and that's one of the things that I think makes downshift so powerful.

So are we going to change the default behavior? No, I don't think that we will. If we did, the people on the other side of the use case spectrum would be frustrated instead. There have to be defaults in downshift, otherwise it wouldn't be very useful. So instead we'll just make changing the defaults easier which is what I think my `keyDownHandlers` prop idea does :)

@hzhu, I'm glad that the solution seems to work for you! I'd love a PR. If you'd prefer to wait for your co-worker to weigh in that's fine with me. I'm not waiting on this issue personally. But I do think the solution is a good one.

**Comment by kentcdodds** [comment]
Ha! I knew that making the solution more generic was a good idea! This solution will also make the arrow key behavior easier to control as well (#336)

**Comment by kentcdodds** [comment]
Oh and @indiesquidge, I forgot to address your "Potential Solution (Change in Downshift Internals)" solution. That wont work because the value of the input and the value of the item is not necessarily the same because the item can be an object and the `itemToString` is responsible for serializing the items to a string for the input value. We don't have a mechanism to deserialize a string value to an item I'm afraid. Consider a list of contacts where two contacts have the same name.

This `keyDownHandlers` idea is what I want to go with. Any takers??? :)

**Comment by austintackaberry** [comment]
I believe @hzhu was taking this one on

**Comment by indiesquidge** [comment]
Hi, @kentcdodds. Those are all super valid points, and I agree that it is not necessarily Downshift's responsibility to implement (or even help provide a way to implement) all of the features for all of the different types of autocompletes, typeaheads, combo boxes, etc.—it could be a never-ending list of requests 😅

I knew I was missing some caveat with my one-off solution, haha. Thanks for explaining why it wouldn't work. Your more generic solution of exposing the keydown handlers seems like a great idea for solving the form submission problem; I can't wait to see it added to this library by @hzhu.

I am very grateful for how much is provided _for free_ for all of us using OSS like this, and it's even more pleasant that you (and other maintainers) are willing to actively maintain and engage in a discussion around the tools you release. Thanks again 💛

**Comment by hzhu** [comment]
> I'm glad that the solution seems to work for you! I'd love a PR. If you'd prefer to wait for your co-worker to weigh in that's fine with me. 

Hi @kentcdodds, my schedule's been hectic but I'll try to submit a PR for review this weekend 😸 !

**Comment by kentcdodds** [comment]
Folks who care, I'd appreciate feedback on the PR: https://github.com/paypal/downshift/pull/346#discussion_r170763621

I'm starting to rethink the decision we've made about this.

**Comment by kentcdodds** [comment]
Alright! So `downshift@1.28.6` now supports [`event.preventDownshiftDefault = true`](https://github.com/paypal/downshift#event-handlers). Much better IMO and will help us avoid this issue entirely :)

Thanks everyone! Especially @ericedem for implementing it and those who contributed to the discussion in #353 :+1: :tada:

**Comment by kentcdodds** [comment]
Hmmm.. I'm not certain what went wrong with the auto-release, but I'm kicking off another one. Hopefully we'll have the actual new version soon.

**Comment by kentcdodds** [comment]
Ok, great! It's out and it's actually `downshift@1.29.0` (because this is a new feature 😅)

Thanks again!
